cmd/maya-apiserver/spc-watcher/lease.go

```diff
@@ -16,29 +16,79 @@ limitations under the License.
 	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
 	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
-	menv "github.com/openebs/maya/pkg/env/v1alpha1"
+	env "github.com/openebs/maya/pkg/env/v1alpha1"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
 	"strings"
 )
 
 const (
 	// SpcLeaseKey is the key that will be used to acquire lease on spc object.
-	// It will be present in spc annotations
+	// It will be present in spc annotations.
 	// If key has an empty value, that means no one has acquired a lease on spc object.
-	SpcLeaseKey = "openebs.io/spc-create-lease"
+	SpcLeaseKey = "openebs.io/spc-lease"
+	// PatchOperation is the strategy of patch operation.
+	PatchOperation = "replace"
+	// PatchPath is the path to the field on spc object which need to be patched.
+	PatchPath = "/metadata/annotations/openebs.io~1spc-lease"
 )
 
+// Patch struct represent the struct used to patch
+// the spc object
+
+// This struct will used to patch the spc object by a lease holder
+// to release the lease once done.
+type Patch struct {
+	// Op defines the operation
+	Op string `json:"op"`
+	// Path defines the key path
+	// eg. for
+	// {
+	//  	"Name": "openebs"
+	//	    Category: {
+	//		  "Inclusive": "v1",
+	//		  "Rank": "A"
+	//	     }
+	// }
+	// The path of 'Inclusive' would be
+	// "/Name/Category/Inclusive"
+	Path  string `json:"path"`
+	Value string `json:"value"`
+}
+
+// This struct will be used as a value of lease key that will
+// give information about an acquired lease on spc
+
+// The struct object will be parsed to string which will be then
+// put as a value to the lease key of spc annotation.
+type lease struct {
+	// Holder is the namespace/name of the pod who acquires the lease
+	Holder           string `json:"holder"`
+	LeaderTransition int    `json:"leaderTransition"`
+	// More specific details can be added here that will describe the
+	// current state of lease in more details.
+	// e.g. acquiredTimeStamp, self-release etc
+	// acquiredTimeStamp will tell when the lease was acquired
+	// self-release will tell whether the lease was removed by the acquirer maya-pod
+	// or by other maya-pod
+}
+
 // Leases is an interface which assists in getting and releasing lease over an spc object
 type Leases interface {
-	// GetLease will try to get a lease on spc, in case of failure it will return error
-	GetLease() (string, error)
-	// UpdateLease will update the lease value of the spc
-	UpdateLease(leaseValue string) (*apis.StoragePoolClaim, error)
-	// RemoveLease will remove the acquired lease on the spc
-	RemoveLease() *apis.StoragePoolClaim
+	// Hold will try to get a lease on spc, in case of failure it will return error
+	Hold() (string, error)
+	// Update will update the lease value of the spc
+	Update(leaseValue string) (*apis.StoragePoolClaim, error)
```

Comment:
Correct. But SPC is returned so that the caller function can extract update value from the object that got updated.
This is helpful in testing whether the expected update was done or not.